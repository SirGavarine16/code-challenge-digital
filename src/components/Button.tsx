import React, { FC } from 'react';
import { FlexStyle, StyleSheet, Text, TouchableOpacity, View } from 'react-native';
import { useTheme } from '@react-navigation/native';

interface Props {
    title?: string;
    onPress?: () => void;
    width?: FlexStyle['width'];
}

const Button: FC<Props> = ({ title = '', onPress, width = '100%' }) => {
    const { colors } = useTheme();
    const { container, text } = styles;

    return (
        <TouchableOpacity activeOpacity={0.8} testID='button' onPress={onPress} style={[container, { backgroundColor: colors.primary, width }]}>
            <Text testID='button-title' style={text}>
                {title}
            </Text>
        </TouchableOpacity>
    );
};

const styles = StyleSheet.create({
    container: {
        padding: 15,
        borderRadius: 10,
    },
    text: {
        color: 'white',
        fontSize: 14,
        fontWeight: '800',
        textAlign: 'center',
    }
})

export default Button;
