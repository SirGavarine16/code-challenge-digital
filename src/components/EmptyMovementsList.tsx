import React, { FC } from 'react';
import { StyleSheet, Text } from 'react-native';

import { colors } from '../themes/AppTheme';

interface Props {
    message?: string;
}

const EmptyMovementsList: FC<Props> = ({ message = 'No hay movimientos disponibles.' }) => {
    return (
        <Text testID='empty-message' style={styles.text}>
            {message}
        </Text>
    );
};

const styles = StyleSheet.create({
    text: {
        fontSize: 16,
        fontWeight: '400',
        textAlign: 'center',
    }
});

export default EmptyMovementsList;
