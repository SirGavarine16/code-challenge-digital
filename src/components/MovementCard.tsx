import React, { FC } from 'react';
import { Image, StyleSheet, Text, TouchableOpacity, View } from 'react-native';
import { useNavigation } from '@react-navigation/native';
import { StackNavigationProp } from '@react-navigation/stack';
import MaterialCommunityIcons from 'react-native-vector-icons/MaterialCommunityIcons';

import { Movement } from '../interfaces';
import { formatDate } from '../utils/formats';
import { AppStackParams } from '../navigation/AppStack';

interface Props {
    item: Movement;
}

const MovementCard: FC<Props> = ({ item }) => {
    const { container, image, dataContainer, boldText, normalText, pointsContainer, pointsText } = styles;

    const { navigate } = useNavigation<StackNavigationProp<AppStackParams>>();

    return (
        <TouchableOpacity testID='card' style={container} onPress={() => navigate('MovementScreen', item)}>
            <Image testID='image' source={{ uri: item.image }} style={image} resizeMode='contain' />
            <View style={dataContainer}>
                <Text testID='name' style={boldText}>
                    {item.product}
                </Text>
                <Text testID='date' style={[normalText, { marginTop: 2 }]}>
                    {formatDate(item.createdAt)}
                </Text>
            </View>
            <View style={pointsContainer}>
                <MaterialCommunityIcons testID='points-icon' name={item.is_redemption ? 'minus' : 'plus'} color={item.is_redemption ? '#FF0000' : '#00B833'} />
                <Text testID='points' style={pointsText}>
                    {item.points}
                </Text>
            </View>
            <MaterialCommunityIcons size={25} name='chevron-right' color='#070707'  />
        </TouchableOpacity>
    );
};

const styles = StyleSheet.create({
    container: {
        flexDirection: 'row',
        alignItems: 'center',
        overflow: 'hidden',
    },
    image: {
        width: 55,
        height: 55,
        resizeMode: 'contain',
        borderRadius: 5,
    },
    dataContainer: {
        marginLeft: 10,
        maxWidth: '50%',
        justifyContent: 'center'
    },
    normalText: {
        fontSize: 12,
        fontWeight: '400',
    },
    boldText: {
        fontWeight: '800'
    },
    pointsContainer: {
        flex: 1,
        flexDirection: 'row',
        alignItems: 'center',
        justifyContent: 'flex-end'
    },
    pointsText: {
        fontSize: 16,
        fontWeight: '800',
    }
});

export default MovementCard;
