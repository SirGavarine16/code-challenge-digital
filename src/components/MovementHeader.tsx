import React, { FC } from 'react';
import { StyleSheet, Text, View } from 'react-native';

import { colors, spacing } from '../themes/AppTheme';

interface Props {
    product?: string;
    top?: number;
}

const MovementHeader: FC<Props> = ({ product = 'Nombre del producto', top = 0 }) => {
    const { container, text } = styles;

    return (
        <View style={[container, { paddingTop: top + 80 }]}>
            <Text testID='product-name' style={text}>
                {product}
            </Text>
        </View>
    );
};

const styles = StyleSheet.create({
    container: {
        backgroundColor: colors.secondary,
        paddingHorizontal: spacing.paddingHorizontal,
        marginBottom: 20,
    },
    text: {
        fontSize: 24,
        fontWeight: '800',
        marginBottom: 10,
    }
})

export default MovementHeader;
