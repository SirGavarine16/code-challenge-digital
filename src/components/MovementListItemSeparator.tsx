import React, { FC } from 'react';
import { StyleSheet, View } from 'react-native';

interface Props {
}

const MovementListItemSeparator: FC<Props> = () => {
    const { container } = styles;
    return (
        <View style={container} />
    );
};

const styles = StyleSheet.create({
    container: {
        height: 5,
    },
});

export default MovementListItemSeparator;
