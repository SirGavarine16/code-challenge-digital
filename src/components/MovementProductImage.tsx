import React, { FC } from 'react';
import { Image, StyleSheet, View } from 'react-native';

import { colors } from '../themes/AppTheme';

interface Props {
    image: string;
}

const MovementProductImage: FC<Props> = ({ image }) => {
    const { container, content } = styles;

    return (
        <View testID='card-image' style={container}>
            <Image testID='product-image' source={{ uri: image }}  style={content} resizeMode='contain' />
        </View>
    );
};

const styles = StyleSheet.create({
    container: {
        backgroundColor: 'white',
        width: '100%',
        aspectRatio: 1 / 1,
        shadowColor: "#000",
        shadowOffset: {
            width: 0,
            height: 4,
        },
        shadowOpacity: 0.30,
        shadowRadius: 4.65,
        elevation: 8,
        borderRadius: 20,
        justifyContent: 'center',
        alignItems: 'center',
    },
    content: {
        width: '100%',
        height: '100%',
        backgroundColor: colors.background
    }
})

export default MovementProductImage;
