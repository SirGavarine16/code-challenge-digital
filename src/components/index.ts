export { default as WelcomeHeader } from './WelcomeHeader';
export { default as PointsDisplay } from './PointsDisplay';
export { default as SectionLabel } from './SectionLabel';