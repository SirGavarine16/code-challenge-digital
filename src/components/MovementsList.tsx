import React, { FC, useCallback } from 'react';
import { StyleSheet, View, FlatList, ListRenderItemInfo } from 'react-native';

import Button from './Button';
import MovementCard from './MovementCard';
import SectionLabel from './SectionLabel';
import MovementListItemSeparator from './MovementListItemSeparator';
import { useFlatListOptimizations, useMovementsListControl } from '../hooks';
import { Movement } from '../interfaces';

interface Props {
    bottom?: number;
}

const MovementsList: FC<Props> = ({ bottom = 0 }) => {
    const { container, listContainer, filtersContainer } = styles;

    const { keyExtractor, getItemLayout } = useFlatListOptimizations<Movement>('id', 55);
    const { data, filter, resetFilter, showNegativeMovements, showPositiveMovements } = useMovementsListControl();

    // Using inline arrow function to render FlatList item re-creates the function on every re-render.
    const renderItem = useCallback(({ item }: ListRenderItemInfo<Movement>) => {
        return <MovementCard item={item} />
    }, [])

    return (
        <View style={[container, { paddingBottom: bottom === 0 ? 20 : 0 }]}>
            <SectionLabel label='Tus movimientos' />
            <View testID='list-container' style={listContainer}>
                <FlatList
                    data={data}
                    renderItem={renderItem}
                    ItemSeparatorComponent={MovementListItemSeparator}
                    keyExtractor={keyExtractor}
                    getItemLayout={getItemLayout}
                    maxToRenderPerBatch={8}
                    showsVerticalScrollIndicator={false}
                />
            </View>
            <View style={filtersContainer}>
                {
                    filter !== ''
                        ? <Button title='Todos' onPress={resetFilter} />
                        : <>
                            <Button title='Ganados' width='48.5%' onPress={showPositiveMovements} />
                            <Button title='Canjeados' width='48.5%' onPress={showNegativeMovements} />
                        </>
                }
            </View>
        </View>
    );
};

const styles = StyleSheet.create({
    container: {
        flex: 1,
    },
    listContainer: {
        backgroundColor: 'white',
        flex: 1,
        borderRadius: 8,
        padding: 10,
    },
    filtersContainer: {
        marginTop: 30,
        flexDirection: 'row',
        justifyContent: 'space-between',
    }
});

export default MovementsList;
