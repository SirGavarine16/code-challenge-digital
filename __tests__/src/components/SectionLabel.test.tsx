import { render } from '@testing-library/react-native';
import { SectionLabel } from '../../../src/components';

describe('<SectionLabel />', () => {
    it('should render correctly', () => {
        const screen = render(<SectionLabel />);
        expect(screen).toMatchSnapshot();
    });

    it('should display the given label', () => {
        const label = 'Testing';
        const { getByText } = render(<SectionLabel label={label} />);
        expect(getByText(label).children[0]).toBe(label);
    });
});