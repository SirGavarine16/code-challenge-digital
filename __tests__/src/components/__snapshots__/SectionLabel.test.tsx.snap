// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SectionLabel /> should render correctly 1`] = `
<Text
  style={
    {
      "color": "#9B9898",
      "fontSize": 14,
      "fontWeight": "800",
      "marginBottom": 20,
      "textTransform": "uppercase",
    }
  }
  testID="section-label"
/>
`;
